### ISSUE TITLE
Searching with Unicode characters 

### ISSUE THREAD

**Issue by fuzunspm** [MM2-0/Kvaesitso#593]
I should able to find contact names or other things named like ABCÜ with a search keyword ABCU but at the moment it doesn't find. Default android can find it

**Comment by leleogere** [comment]
Yes! A search that do not take accents into accounts is a must have for a lot of languages

**Comment by Sir-Photch** [comment]
Do I understand correctly, that search terms like "cafe" should match both "café" and "cáfè"? (ignoring that this is not a word, for a second)

In this case, you want to use ASCII-Characters to match for non-ASCII (Unicode). Do I understand correctly?

**Comment by leleogere** [comment]
For now, it seems to be working for apps (searching `velib` does bring up the app `vélib'`), at least with simple accents like "é". But it does not work for contacts (and maybe others things?), for example searching for `Leo` will not show up a contact named `Léo`.

**Comment by Sir-Photch** [comment]
For contacts, Kvaesitso is querying Android directly with the search term, and looking at the source code I don't see a way to tell the query to romanize possible results first.

You can see the implementation for contacts here:
https://github.com/MM2-0/Kvaesitso/blob/4c0b34e091e0779a164fd60eeaa0941dcb2da916/data/contacts/src/main/java/de/mm20/launcher2/contacts/ContactRepository.kt#L174-L200

**Comment by leleogere** [comment]
I'm not familiar at all with Android unfortunately, but there must be some way to do it as [Sesame search](https://play.google.com/store/apps/details?id=ninja.sesame.app.edge&hl=fr&gl=US&pli=1) is able to do it

**Comment by fuzunspm** [comment]
> Do I understand correctly, that search terms like "cafe" should match both "café" and "cáfè"? (ignoring that this is not a word, for a second)
> 
> In this case, you want to use ASCII-Characters to match for non-ASCII (Unicode), right?
> 
> There is a library that implements this: https://github.com/anyascii/anyascii
> 
> This would imply that every app name also needs to be stored in it's romanized form, which then could be checked when searching.

yes it should match both. Other examples:

sirket > şirket or sirket
uzum > üzüm or uzum
ağrı > agri or ağrı or agrı or ağri

**Comment by MM2-0** [comment]
> For contacts, Kvaesitso is querying Android directly with the search term, and looking at the source code I don't see a way to tell the query to romanize possible results first.

Yep, that's pretty much the reason. For apps and shortcuts, we keep every single app and shortcut in memory so we are more flexible and therefore we can strip accents when we filter the results, but for contacts, calendar events, and local files, we query Android's database directly which unfortunately only supports substring queries. I admit, it's not ideal.

**Comment by Sir-Photch** [comment]
There are a lot of issues open that complain about performance, and caching everything may add other hiccups as well as increase memory usage. Also, this cache then needs to be kept up-to-date when there are changes to contacts. 

But with that cache, it should be possible.

**Comment by MM2-0** [comment]
> Wouldn't keeping contacts in memory like apps be possible?

I guess for contacts it's possible. The problem is that it doesn't scale very well. But I guess most people don't have more than 500 different contacts, so that would work. It's more problematic with files and calendar events.

What we could do is query 1000 files, filter them, query the next 1000 files, filter them, … repeat until we have the desired number of search results or until we have searched all files. But that would be significantly slower than just a single query. 

Third option would be to copy the entire database to the launcher's internal database, and add some additional search index columns. But then we'd need to make sure that the data is always in sync and it's just a lot of additional maintenance cost for little to no gain (usually I know how my contacts are spelt).

**Comment by leleogere** [comment]
The issue for me is mainly about the contacts as that's what I use the most. It would be handy for files (and probably events, I don't know I don't use my launcher for that), but I understand that it is not wouldn't be easy to implement and that it can be kinda slow. 

> (usually I know how my contacts are spelt).

Well, I usually know how my contacts are spelt too. The thing is that I see this type of search-based launcher as something that allow to get very quickly to what I'm looking for, and while I can type `leo` in less than half a second, the time taken to type `léo` is longer than a second as it is limited by the time the accent pop-up of my keyboard takes to appear when I long press the "e". And it can be even longer in case of multiple accents like `Éléonore`.
